pkg/ruler/manager_test.go

```diff
@@ -153,30 +121,95 @@ func getManager(m *DefaultMultiTenantManager, user string) RulesManager {
 
 	return m.userManagers[user]
 }
+func assertManagerMockRunningForUser(t *testing.T, m *DefaultMultiTenantManager, userID string) *managerMock {
+	t.Helper()
+
+	rm := getManager(m, userID)
+	require.NotNil(t, rm)
+
+	// The ruler manager start is async, so we poll it.
+	test.Poll(t, 1*time.Second, true, func() interface{} {
+		return rm.(*managerMock).running.Load()
+	})
+
+	return rm.(*managerMock)
+}
+
+func assertManagerMockNotRunningForUser(t *testing.T, m *DefaultMultiTenantManager, userID string) {
+	t.Helper()
+
+	rm := getManager(m, userID)
+	require.Nil(t, rm)
+}
+
+func assertManagerMockStopped(t *testing.T, m *managerMock) {
+	t.Helper()
+
+	// The ruler manager stop is async, so we poll it.
+	test.Poll(t, 1*time.Second, false, func() interface{} {
+		return m.running.Load()
+	})
+}
+
+func assertRuleGroupsMappedOnDisk(t *testing.T, m *DefaultMultiTenantManager, userID string, expectedRuleGroups rulespb.RuleGroupList) {
+	t.Helper()
+
+	// Verify that the rule groups have been mapped on disk for the given user.
+	users, err := m.mapper.users()
+	require.NoError(t, err)
+
+	if len(expectedRuleGroups) > 0 {
+		require.Contains(t, users, userID)
+	} else {
+		require.NotContains(t, users, userID)
+	}
+
+	// Verify the content of the rule groups mapped to disk.
+	for namespace, expectedFormattedRuleGroups := range expectedRuleGroups.Formatted() {
+		// The mapper sort groups by name in reverse order, so we apply the same sorting
+		// here to expected groups.
+		sort.Slice(expectedFormattedRuleGroups, func(i, j int) bool {
+			return expectedFormattedRuleGroups[i].Name > expectedFormattedRuleGroups[j].Name
+		})
+
+		expectedYAML, err := yaml.Marshal(rulefmt.RuleGroups{Groups: expectedFormattedRuleGroups})
+		require.NoError(t, err)
+
+		path := filepath.Join(m.mapper.Path, userID, namespace)
+		file, err := m.mapper.FS.Open(path)
+		require.NoError(t, err)
+
+		content, err := io.ReadAll(file)
+		require.NoError(t, err)
+		assert.Equal(t, string(expectedYAML), string(content))
+
+		require.NoError(t, file.Close())
+	}
+}
 
-func factory(_ context.Context, _ string, _ *notifier.Manager, _ log.Logger, _ prometheus.Registerer) RulesManager {
-	return &mockRulesManager{done: make(chan struct{})}
+func managerMockFactory(_ context.Context, _ string, _ *notifier.Manager, _ log.Logger, _ prometheus.Registerer) RulesManager {
+	return &managerMock{done: make(chan struct{})}
 }
 
-type mockRulesManager struct {
+type managerMock struct {
 	running atomic.Bool
 	done    chan struct{}
 }
 
-func (m *mockRulesManager) Run() {
+func (m *managerMock) Run() {
 	m.running.Store(true)
 	<-m.done
 }
 
-func (m *mockRulesManager) Stop() {
+func (m *managerMock) Stop() {
 	m.running.Store(false)
```

Comment:
Shouldn't we call `m.running.Store(false)` via `defer` in `Run`, when it actually stops running? (Now checking for `running == false` only tests that `Stop` was called.)

Reply:
Yes, done here: https://github.com/grafana/mimir/pull/4994/commits/cd0d725a46e24e67a78f62990d85397f076196a6